std/arith.asm

```diff
@@ -140,25 +304,25 @@ machine Arith(CLK32_31, operation_id){
     *
     *******/
 
-    // 0xffffffffffffffffffffffffffffffffffffffffffffffffffff fffe ffff fc2f
-    let p = array_as_fun([
-        0xfc2f, 0xffff, 0xfffe, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff,
-        0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff
-    ]);
+    let p: int -> int = |i| if i >= 0 && i < 16 {
+        (secp_modulus >> (i * 16)) & 0xffff
+    } else {
+        0
+    };
 
     // The "- 4 * shift_right(p, 16)" effectively subtracts 4 * (p << 16 * 16) = 2 ** 258 * p
     // As a result, the term computes `(x - 2 ** 258) * p`.
-    let product_with_p = (|| |x| |nr| product(p, x)(nr) - 4 * shift_right(p, 16)(nr))();
+    let product_with_p: int -> (int -> expr) = |x| |nr| product(p, x)(nr) - 4 * shift_right(p, 16)(nr);
 
-    let eq1 = (|| |nr| product(sf, x2f)(nr) - product(sf, x1f)(nr) - y2f(nr) + y1f(nr) + product_with_p(q0f)(nr))();
+    let eq1: int -> expr = |nr| product(sf, x2f)(nr) - product(sf, x1f)(nr) - y2f(nr) + y1f(nr) + product_with_p(q0f)(nr);
 
     /*******
     *
     * EQ2:  2 * s * y1 - 3 * x1 * x1 + (q0 * p)
     *
     *******/
 
-    let eq2 = (|| |nr| 2 * product(sf, y1f)(nr) - 3 * product(x1f, x1f)(nr) + product_with_p(q0f)(nr))();
+    let eq2: int -> expr = |nr| 2 * product(sf, y1f)(nr) - 3 * product(x1f, x1f)(nr) + product_with_p(q0f)(nr);
```

Comment:
I'm not 100% sure about these types, but it seems to work and let's use remove the hack.